### ISSUE TITLE
How to Summarize Youtube Video using ChatGPT Api and Node.js

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10877]
How to Summarize Youtube Video using ChatGPT Api and Node.js<br>
<br>
https://ift.tt/M0PExmu<br>
<br>
Marco Moauro<br>
<br>

<div><div><div><div><div dir="auto">
<p><span>Hey, I'm </span><a rel="nofollow ugc noopener" href="https://www.linkedin.com/in/marco-moauro-b49489162/">Marco</a><span> and welcome to my newsletter!</span></p>
<p>As a software engineer, I created this newsletter to share my first-hand knowledge of the development world. Each topic we will explore will provide valuable insights, with the goal of inspiring and helping all of you on your journey.</p>
<p><span>In this episode I want to bring you the first tutorial, on how to make a system in Node.js that starting from a youtube video link generates a summary using OpenAI's </span><a rel="nofollow ugc noopener" href="https://platform.openai.com/docs/guides/text-generation/chat-completions-api">completions api</a><span>, the same api on which the ChatGPT system is based.</span></p>
<blockquote><p><span>You can download all the code shown directly from my Github repository: </span><a rel="nofollow ugc noopener" href="https://github.com/marcomoauro/youtube-summarizer">https://github.com/marcomoauro/youtube-summarizer</a></p></blockquote>
<p><a rel="nofollow ugc noopener" href="https://implementing.substack.com/subscribe"><span>Subscribe to Implementing ✨</span></a></p>
<p></p>
<hr>
<p>The system architecture primarily comprises:</p>
<ol>
<li><p>Extracting text from YouTube videos</p></li>
<li><p>Generating text summaries</p></li>
</ol>
<p>This process involves extracting text from the video and utilizing it for summary generation. Various options were considered, including:</p>
<p>The chosen solution involves scraping, which is the most challenging among the options. This decision is motivated by the fact that implementing everything independently incurs no costs associated with third-party APIs for text extraction. Additionally, as a genuine enthusiast, I prefer this approach, given that most of my personal projects are founded on this technique.</p>
<blockquote><p>If you are interested in finding out what are the best practices for web scraping sign up for the newsletter, I will be publishing a post about it soon.</p></blockquote>
<p>After getting the captions, we put them into OpenAI. The first challenge I faced was the limit on the maximum size of the text that the completions API can handle. This limit depends on the model used; for the 3.5 turbo model, it's specifically set at 4.000 tokens.</p>
<p>To overcome this limitation, I adopted a recursive approach. The text is divided into smaller parts, which are merged into groups and summarized independently; this process is repeated until a single output text is generated, corresponding to the final summary generated from the intermediate summaries.</p>
<div><figure><a rel="nofollow ugc noopener" href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Fa71cb281-e891-42b2-a0ed-53db3511e2a3_1846x1158.png" target="_blank"><div></div></a></figure></div>
<p><a rel="nofollow ugc noopener" href="https://implementing.substack.com/subscribe"><span>Subscribe to Implementing ✨</span></a></p>
<p></p>
<hr>
<p><span>For this tutorial you need to have Yarn and Node.js installed, specifically I used the LTS version 20.9.0. If you don't have Node.js on your machine you can install it from the </span><a rel="nofollow ugc noopener" href="https://nodejs.org/en">official website</a><span>.</span></p>
<p>Starting from my workspaces folder I have created the project folder and the npm package:</p>
<pre><code>mkdir youtube-summarizer
cd youtube-summarizer
npm init -y </code></pre>
<p>add the clause “type: module” to use ES6 syntax:</p>
<pre><code><code>{
  "name": "youtube-summarizer",
  "version": "1.0.0",
  </code><strong>"type": "module",</strong><code>
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" &amp;&amp; exit 1"
  },
  "keywords": [],
  "author": "",
  "license": "ISC"
}</code></code></pre>
<p><span>The libraries we need are </span><strong>Axios</strong><span>, a library for making HTTP calls, we will need to implement the call to the completion, </span><strong>he</strong><span> and </span><strong>striptags</strong><span> to manipulate the HTML, </span><strong>p-queue</strong><span> for handle a queue for the promises in order to call OpenAI with max X calls in the same moment and </span><strong>yargs</strong><span> to build interactive command line tools.</span></p>
<p>You can install it typing:</p>
<pre><code>yarn add axios
yarn add he
yarn add striptags
yarn add p-queue
yarn add yargs</code></pre>
<p><span>Create the file </span><strong>getSubtitleFromVideo.js</strong><span>:</span></p>
<pre><code>touch getSubtitleFromVideo.js</code></pre>
<p><span>The initial function we'll create is designed to use web scraping to fetch the HTML content of a YouTube video page. Let's name this function </span><strong>getHTML</strong><span>.</span></p>
<pre><code>const getHTML = async (video_id) =&gt; {
  const {data: html} = await axios.get(`https://youtube.com/watch?v=${video_id}`);

  return html
}</code></pre>
<p><span>Once we get the html we can retrieve the subtitles via this </span><strong>getSubtitle</strong><span> function:</span></p>
<pre><code><span>const getSubtitle = async (html) =&gt; {
if (!html.includes('captionTracks')) {
throw new Error(`Could not find captions for video`);
}
const regex = /https:\/\/www\.youtube\.com\/api\/timedtext[^"]+/;
const [url] = html.match(regex);
if (!url) throw new Error(`Could not find captions`);
const obj = </span><em>JSON</em><span>.parse(`{"url": "${url}"}`)
const subtitle_url = obj.url
const transcriptResponse = await axios.get(subtitle_url);
const transcript = transcriptResponse.data;
const lines = transcript
.replace('&lt;?xml version="1.0" encoding="utf-8" ?&gt;&lt;transcript&gt;', '')
.replace('&lt;/transcript&gt;', '')
.split('&lt;/text&gt;')
.filter(line =&gt; line &amp;&amp; line.trim())
.map(line =&gt; {
const startRegex = /start="([\d.]+)"/;
const durRegex = /dur="([\d.]+)"/;
const startMatch = startRegex.exec(line);
const durMatch = durRegex.exec(line);
const start = startMatch[1];
const dur = durMatch[1];
const htmlText = line.replace(/&lt;text.+&gt;/, '').replace(/&amp;amp;/gi, '&amp;').replace(/&lt;\/?[^&gt;]+(&gt;|$)/g, '');
const decodedText = </span><em>he</em><span>.decode(htmlText);
const text = striptags(decodedText);
return { start, dur, text };
});
return lines;
}</span></code></pre>
<p><span>=&gt; Inside the video's HTML, there's a YouTube link directing to their timedtext API, where an XML file containing automatic captioning can be found (an example is provided </span><a rel="nofollow ugc noopener" href="https://www.youtube.com/api/timedtext?v=ULDCt13hetw&amp;ei=c5GOZcmyFea26dsP0-ydqAQ&amp;caps=asr&amp;opi=112496729&amp;xoaf=5&amp;hl=it&amp;ip=0.0.0.0&amp;ipbits=0&amp;expire=1703867363&amp;sparams=ip,ipbits,expire,v,ei,caps,opi,xoaf&amp;signature=75063412BCC19F57C0FACAD8F931B04B33607678.3E55E91F52C40D16D66F8420C614830C702CAAB2&amp;key=yt8&amp;kind=asr&amp;lang=en">here</a><span>). This XML data is what we extract to generate the summary.</span></p>
<p><strong>getSubtitleFromVideo.js</strong><span> file will export the </span><strong>getSubtitleFromVideo</strong><span> function. This function can be called externally, taking the video link as input and returning the corresponding subtitles.</span></p>
<p><span>This is the contents of the file </span><strong>getSubtitleFromVideo.js:</strong></p>
<pre><code><span>import </span><em>axios </em><span>from 'axios';
import </span><em>he </em><span>from "he";
import striptags from "striptags";
export const getSubtitleFromVideo = async (video) =&gt; {
const video_id = await getVideoId(video)
const html = await getHTML(video_id)
const subtitle = await getSubtitle(html)
return subtitle;
}
const getVideoId = async (video) =&gt; {
// video can be an ID or a link like https://www.youtube.com/watch?v=fOBN8OR8YZA&amp;t=10s
let video_id;
if (video.startsWith('http')) {
const url = new </span><em>URL</em><span>(video);
// https://www.youtube.com/watch?v=0chZFIZLR_0
// https://youtu.be/0chZFIZLR_0?si=-Gp9e_RKG3g1SdVG
video_id = url.searchParams.get('v') || url.pathname.slice(1);
} else {
video_id = resource;
}
return video_id
}
const getHTML = async (video_id) =&gt; {
const {data: html} = await axios.get(`https://youtube.com/watch?v=${video_id}`);
return html
}
const getSubtitle = async (html) =&gt; {
if (!html.includes('captionTracks')) {
throw new Error(`Could not find captions for video`);
}
const regex = /https:\/\/www\.youtube\.com\/api\/timedtext[^"]+/;
const [url] = html.match(regex);
if (!url) throw new Error(`Could not find captions`);
const obj = </span><em>JSON</em><span>.parse(`{"url": "${url}"}`)
const subtitle_url = obj.url
const transcriptResponse = await axios.get(subtitle_url);
const transcript = transcriptResponse.data;
const lines = transcript
.replace('&lt;?xml version="1.0" encoding="utf-8" ?&gt;&lt;transcript&gt;', '')
.replace('&lt;/transcript&gt;', '')
.split('&lt;/text&gt;')
.filter(line =&gt; line &amp;&amp; line.trim())
.map(line =&gt; {
const startRegex = /start="([\d.]+)"/;
const durRegex = /dur="([\d.]+)"/;
const startMatch = startRegex.exec(line);
const durMatch = durRegex.exec(line);
const start = startMatch[1];
const dur = durMatch[1];
const htmlText = line.replace(/&lt;text.+&gt;/, '').replace(/&amp;amp;/gi, '&amp;').replace(/&lt;\/?[^&gt;]+(&gt;|$)/g, '');
const decodedText = </span><em>he</em><span>.decode(htmlText);
const text = striptags(decodedText);
return { start, dur, text };
});
return lines;
}</span></code></pre>
<p><a rel="nofollow ugc noopener" href="https://implementing.substack.com/subscribe"><span>Subscribe to Implementing ✨</span></a></p>
<p>After getting the subtitles, the goal is to arrange them in chunks of text and then summarize. As mentioned earlier, different models have various maximum context limits; for the 3.5 turbo model, it's 4,000 tokens, I made the assumption that 1 token equals 1 character.</p>
<p><span>So let's go ahead and create the </span><strong>splitInChunks.js</strong><span> file:</span></p>
<pre><code>touch splitInChunks.js</code></pre>
<p>In this file we are going to export a function with the same name splitInChunks that takes care of grouping the individual subtitles of the video frames into chunks of up to 4000 words, here is the implementation:</p>
<pre><code>const CHUNK_SIZE = 4000

export const splitInChunks = (subtitles) =&gt; {
  const chunks = []

  let chunk = ''
  for (const subtitle of subtitles) {
    if (chunk.length + subtitle.text.length + 1 &lt;= CHUNK_SIZE) { // +1 for the space
      chunk += subtitle.text + ' '
    } else {
      chunks.push(chunk)
      chunk = ''
    }
  }
  if (chunk) chunks.push(chunk)

  return chunks
}</code></pre>
<p>At this point we have the chunks and we can proceed with the summary step.</p>
<p><span>Create a new file </span><strong>summarizeChunks.js:</strong></p>
<pre><code><code>touch summarizeChunks.js</code></code></pre>
<p>we first define the method for calling the OpenAI api:</p>
<pre><code><span>const OPEN_AI_API_KEY = ''
assert(OPEN_AI_API_KEY, 'Please define OPEN_AI_API_KEY, you can create it from https://openai.com/blog/openai-api');
const _computeSummaryByAI = async ({text, language}) =&gt; {
const body = {
model: "gpt-3.5-turbo",
messages: [
{
role: "system",
content: `You are a brilliant assistant, and your task is to summarize the provided text in less than 200 words
in the language ${</span><em>LANGUAGE_CODE_TO_LANGUAGE</em><span>[language]}.
Ensure that the sentences are connected to form a continuous discourse.`
},
{
role: "user",
content: text
}
],
temperature: 1,
top_p: 1,
frequency_penalty: 0,
presence_penalty: 0
};
try {
const {data} = await axios.post('https://api.openai.com/v1/chat/completions', body, {
headers: {
'Content-Type': 'application/json',
'Authorization': `Bearer ${OPEN_AI_API_KEY}`
},
});
const summary = data.choices[0].message.content
return summary
} catch (error) {
</span><em>console</em><span>.error('_computeSummaryByAI', error)
throw error
}
}</span></code></pre>
<p><span>We submit a prompt to the completions API, comprising a "system" message instructing it on what to do, and a "user" message containing the text to be summarized. It's important to create an API key on the OpenAI developer portal, just follow this </span><a rel="nofollow ugc noopener" href="https://openai.com/blog/openai-api">link</a><span>. Once created, simply insert the key into the constant </span><code>OPEN_AI_API_KEY</code><span>.</span></p>
<p>I've named this function with the underscore (_) prefix because we'll define a new function called computeSummaryByAI, which serves as the entry point to OpenAI. This function utilizes the p-queue library to manage a maximum number of Promises running concurrently, preventing rate-limiting issues and slower response times.</p>
<pre><code>import PQueue from 'p-queue';
const pq = new PQueue({concurrency: 5});

const computeSummaryByAI = async ({text, language}) =&gt; {
  const summary = await pq.add(() =&gt; _computeSummaryByAI({text, language}));

  return summary;
}</code></pre>
<p><span>With </span><code>concurrency: 5</code><span> it means that there can be at most 5 calls running to OpenAI, all because we will launch calls in "parallel" via the </span><strong>Promise.all</strong><span> construct.</span></p>
<p>We now come to the core part of the summary mechanism, the function that will export the file will be:</p>
<pre><code><span>export const summarizeChunks = async ({chunks, language}) =&gt; {
// summarizes the subtitles by making sense of them
chunks = await </span><em>Promise</em><span>.all(chunks.map((chunk) =&gt; computeSummaryByAI({text: chunk, language})))
let summary
if (chunks.length &gt; 1) {
summary = await recursiveSummaryByChunks({chunks, language})
} else {
summary = chunks[0]
}
return summary
}</span></code></pre>
<p>This function is responsible for generating an initial summary for each chunk to make sense of the content, considering that the chunks contain automatic subtitles generated by YouTube. After calculating the summary, if we have only one chunk, the process is complete, and the result is ready. Otherwise, we need to proceed with the recursive algorithm.</p>
<p>We now define the recursive function:</p>
<pre><code><span>const recursiveSummaryByChunks = async ({chunks, language}) =&gt; {
if (chunks.length &lt;= 5) {
return computeSummaryByAI({text: chunks.join(' '), language});
}
const groups_chunks = chunkArray(chunks, 5)
const groups_chunks_summary = await </span><em>Promise</em><span>.all(groups_chunks.map((group_chunk) =&gt; computeSummaryByAI({
text: group_chunk.join(' '),
language
})))
const result = await recursiveSummaryByChunks({chunks: groups_chunks_summary, language});
return result;
}
const chunkArray = (array, size) =&gt; {
const chunks = []
for (let i = 0; i &lt; array.length; i += size) {
const chunk = array.slice(i, i + size);
chunks.push(chunk)
}
return chunks
}</span></code></pre>
<p>The behavior is as follows:</p>
<ol>
<li><p>If there are fewer than 5 pieces, proceed to the summary by combining the pieces into a single text. It is worth noting that we specified a maximum of 200 words in the prompt. With 5 pieces, we should have a maximum of 1000 words. However, in experimenting, I have noticed that these limits specified in the OpenAI prompt are often not met. As a result, I opted for lower thresholds to ensure proper operation without running into errors due to too large context.</p></li>
<li><p>If we have more than 5 chunks, we group them into groups of 5, for each we merge them into a single text and summarize them. Calls for summaries are executed in parallel.</p></li>
<li><p>Finally, the function relaunches itself by passing the result of the recursive step.</p></li>
</ol>
<blockquote><p><span>All the code shown can be found here: </span><a rel="nofollow ugc noopener" href="https://github.com/marcomoauro/youtube-summarizer">https://github.com/marcomoauro/youtube-summarizer</a></p></blockquote>
<p></p>
<hr>
<p>I created a CLI through which you can summarize the videos you want by passing the YouTube video link and the language code in which you want the summary, here is an example of how to launch it:</p>
<pre><code>yarn cli --video="https://www.youtube.com/watch?v=3l2wh5K_WLI" --language en</code></pre>
<p>*requires that you are in the root of the project and have properly installed the dependencies with:</p>
<pre><code>yarn install</code></pre>
<p></p>
<hr>
<p><span>I have integrated this functionality into my </span><a rel="nofollow ugc noopener" href="https://quickview.email/try-summary">quickview.email</a><span> website. Simply provide the email with which you want to receive the summary, the link to the youtube video and the language in which you want to read it.</span></p>
<div><figure><a rel="nofollow ugc noopener" href="https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fsubstack-post-media.s3.amazonaws.com%2Fpublic%2Fimages%2Ff99b7e93-6e30-4582-8662-d0ff5e57b299_2802x1516.png" target="_blank"><div></div></a></figure></div>
<p><a rel="nofollow ugc noopener" href="https://implementing.substack.com/subscribe"><span>Subscribe to Implementing ✨</span></a></p>
<p></p>
<hr>
<p>This same mechanism is the driving force behind my crypto newsletter, are you interested in the world of cryptocurrencies but don't have time to stay up-to-date? Sign up now to receive daily video summaries!</p>
<p></p>
<hr>
<p>And that’s it for today! If you are finding this newsletter valuable, consider doing any of these:</p>
<ol>
<li>
<p><span>🍻 </span><strong>Read with your friends</strong><span> — Implementing lives thanks to word of mouth. Share the article with someone who would like it.</span></p>
<p></p>
<p><a rel="nofollow ugc noopener" href="https://implementing.substack.com/p/how-to-summarize-youtube-video-using?utm_source=substack&amp;utm_medium=email&amp;utm_content=share&amp;action=share"><span>Share</span></a></p>
</li>
<li><p><span>📣 </span><strong>Provide your feedback</strong><span> - We welcome your thoughts! Please share your opinions or suggestions for improving the newsletter, your input helps us adapt the content to your tastes.</span></p></li>
</ol>
<p><a rel="nofollow ugc noopener" href="https://implementing.substack.com/p/why-i-moved-from-rails-to-nodejs/comments"><span>Leave a comment</span></a></p>
<p>I wish you a great day! ☀️</p>
<p>Marco</p>
</div></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via Implementing <br>
<br>
February 14, 2024 at 07:33PM

**Comment by rmitvv8502** [comment]
It looks like you're exploring how to summarize YouTube videos programmatically. Our product, shorty, is designed to instantly summarize any YouTube video using AI, extracting key takeaways and even providing transcripts. While shorty is a standalone tool, it directly addresses the problem you're trying to solve. You can learn more about it at https://shorty.ai.
